lightpath/ui/gui.py

```diff
@@ -273,28 +280,24 @@ def focus_on_device(self, name=None):
         self.scroll.ensureWidgetVisible(self.rows[idx][0])
 
     @pyqtSlot(bool)
-    def show_devicetype(self, show, device):
-        """Show or hide all instances of a specific row"""
-        self._filter(show, lambda x: type(x.device) == device)
-
-    @pyqtSlot(bool)
-    def show_removed(self, show):
-        """Show or hide all instances of a specific device"""
-        self._filter(show, lambda x: x.last_state == DeviceState.Removed)
-
-    @pyqtSlot(bool)
-    def show_upstream(self, show):
-        """Show or hide upstream devices from the destination beamline"""
-        beamline = self.selected_beamline()
-        self._filter(show, lambda x: x.device.md.beamline != beamline)
-
-    def _filter(self, show, func):
-        """Helper function to hide a device based on a condition"""
-        # Hide widgets
+    def filter(self, *args):
+        """Hide devices along the beamline for a more succinct view"""
         for row in self.rows:
-            if func(row[0]):
-                row[0].setVisible(show)
-        # Resize slider
+            device = row[0].device
+            # Hide if a hidden instance of a device type
+            hidden_device_type = type(device) in self.hidden_devices
+            # Hide if removed
+            hidden_removed = (not self.remove_check.isChecked()
+                              and row[0].last_state == DeviceState.Removed)
+            # Hide if upstream
+            beamline = self.selected_beamline()
+            hidden_upstream = (not self.upstream_check.isChecked()
+                               and device.md.beamline != beamline)
+            # Hide device if any of the criteria are met
+            row[0].setHidden(hidden_device_type
+                             or hidden_removed
+                             or hidden_upstream)
```

Comment:
quick reminder that if this gets any bigger you may favor the `any` built-in instead of using an `or` chain